### PR TITLE
cleanup code to reduce duplication

### PR DESCRIPTION
**Pull request by mengqiy**
This change is [Reviewable](https://reviewable.io/reviews/redborian/go-types-to-jsonschema/10)

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -77,188 +74,107 @@ func extractFromTag(tag *ast.BasicLit) (string, string) {
...
 	def := &v1beta1.JSONSchemaProps{}
 	externalTypeRefs := []TypeReference{}
+	var typeName, pkgAlias string
 
 	def.Description = comment
 
 	switch tt := t.(type) {
 	case *ast.Ident:
-		typeName := tt.Name
-		if isSimpleType(typeName) {
```
Comment (mengqiy) [inline main.go:89]: This is handled outside of the switch.

File: main.go [hunk 2]
```diff
@@ -77,188 +74,107 @@ func extractFromTag(tag *ast.BasicLit) (string, string) {
...
-			def.Type = jsonifyType(typeName)
-		} else {
-			def.Ref = getPrefixedDefLink(typeName, curPkgPrefix)
-		}
+		typeName = tt.Name
 	case *ast.ArrayType:
-		arrayType := tt
-		var typeNameOfArray, packageAlias string
```
Comment (mengqiy) [inline main.go:96]: `ArrayType` are handled by a recursion in `arrayTypeToSchema`.

File: main.go [hunk 3]
```diff
@@ -77,188 +74,107 @@ func extractFromTag(tag *ast.BasicLit) (string, string) {
...
+	case *ast.SelectorExpr:
+		selectorType := tt
+		pkgAlias = selectorType.X.(*ast.Ident).Name
+		typeName = selectorType.Sel.Name
 
-		if def.Items == nil {
-			def.Items = &v1beta1.JSONSchemaPropsOrArray{}
-		}
-		def.Items.Schema = new(v1beta1.JSONSchemaProps)
-		if isSimpleType(typeNameOfArray) {
```
Comment (mengqiy) [inline main.go:117]: This is handled outside of the switch.

File: main.go [hunk 4]
```diff
@@ -77,188 +74,107 @@ func extractFromTag(tag *ast.BasicLit) (string, string) {
...
-				def.Items.Schema.Ref = getPrefixedDefLink(typeNameOfArray, curPkgPrefix)
-			}
+			def.Ref = getPrefixedDefLink(typeName, curPkgPrefix)
 		}
-		def.Type = "array"
+	}
 
-	case *ast.MapType:
-		mapType := tt
-		switch mapType.Value.(type) {
```
Comment (mengqiy) [inline main.go:130]: `MapType` is extracted into `mapTypeToSchema`

File: main.go [hunk 5]
```diff
@@ -77,188 +74,107 @@ func extractFromTag(tag *ast.BasicLit) (string, string) {
...
+		Items:       &v1beta1.JSONSchemaPropsOrArray{Schema: items},
+		Description: comment,
+	}
+	return props, extRefs
+}
 
-		def.Ref = getPrefixedDefLink(typeName, importPaths[packageAlias])
-		externalTypeRefs = append(externalTypeRefs, TypeReference{typeName, packageAlias})
-	case *ast.StarExpr:
-		starExpr := tt
```
Comment (mengqiy) [inline main.go:156]: `StarExpr` is handled by a recursion.

File: main.go [hunk 6]
```diff
@@ -77,188 +74,107 @@ func extractFromTag(tag *ast.BasicLit) (string, string) {
...
 		} else {
-			if packageAlias != "" {
-				def.Ref = getPrefixedDefLink(typeName, importPaths[packageAlias])
-			} else {
-				def.Ref = getPrefixedDefLink(typeName, curPkgPrefix)
-			}
+			def.AdditionalProperties.Schema.Ref = getPrefixedDefLink(valueType.Name, curPkgPrefix)
 		}
-	case *ast.StructType:
-		structType := tt
```
Comment (mengqiy) [inline main.go:180]: `StructType` is handled by recursions in `structTypeToSchema`.

File: main.go [hunk 7]
```diff
@@ -77,188 +74,107 @@ func extractFromTag(tag *ast.BasicLit) (string, string) {
...
-				inlineDefinitions = append(inlineDefinitions, inlinedDef)
-				// def.AnyOf = append(def.AnyOf, v1beta1.JSONSchemaProps{Ref: getPrefixedDefLink(typeName, curPkgPrefix)})
-				continue
-			}
+	case *ast.InterfaceType:
+		// No op
+		panic("Map Interface Type")
+	}
+	def.Type = "object"
+}
```
Comment (redborian) [inline main.go:150]: Comment about this would be helpful.

File: main.go [hunk 8]
```diff
@@ -77,188 +74,107 @@ func extractFromTag(tag *ast.BasicLit) (string, string) {
...
-		}
-		def.Items.Schema = new(v1beta1.JSONSchemaProps)
-		if isSimpleType(typeNameOfArray) {
-			def.Items.Schema.Type = jsonifyType(typeNameOfArray)
+		def.Ref = getPrefixedDefLink(typeName, importPaths[pkgAlias])
+		externalTypeRefs = []TypeReference{{TypeName: typeName, PackageName: pkgAlias}}
+	case *ast.StarExpr:
+		def, externalTypeRefs = exprToSchema(tt.X, comment, importPaths, curPkgPrefix)
+	case *ast.StructType:
+		externalTypeRefs, def = structTypeToSchema(tt, importPaths, curPkgPrefix, def)
```
Comment (redborian) [inline main.go]: For consistency, (at least for now) can we stick to `def, externalTypeRefs` as written above?

**Review by redborian — APPROVED** [review]
Overall, the change looks good. Just a few comments. Feel free to merge it after you resolve them.

**Comment by mengqiy** [comment]
Addressed comments.
Merging

**Pull request merged**